Source/Core/CustomProjection.js

```diff
@@ -0,0 +1,311 @@
+define([
+        './Cartesian3',
+        './Cartographic',
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties',
+        './DeveloperError',
+        './Ellipsoid',
+        './getAbsoluteUri',
+        './isDataUri',
+        './loadAndExecuteScript',
+        '../ThirdParty/when'
+    ], function(
+        Cartesian3,
+        Cartographic,
+        Check,
+        defaultValue,
+        defined,
+        defineProperties,
+        DeveloperError,
+        Ellipsoid,
+        getAbsoluteUri,
+        isDataUri,
+        loadAndExecuteScript,
+        when) {
+    'use strict';
+
+    /**
+     * {@link MapProjection} that uses custom project and unproject functions defined in an external file.
+     *
+     * The external file must contain a function named <code>createProjectionFunctions</code> that implements the
+     * <code>CustomProjection~factory</code> interface to provide <code>CustomProjection~project</code> and
+     * <code>CustomProjection~unproject</code> functions to a callback.
+     *
+     * Scenes using CustomProjection will default to <code>MapMode2D.ROTATE</code> instead of <code>MapMode2D.INFINITE_SCROLL</code>.
+     *
+     * @alias CustomProjection
+     * @constructor
+     *
+     * @param {String} url The url of the external file.
+     * @param {Ellipsoid} [ellipsoid=Ellipsoid.WGS84] The MapProjection's ellipsoid.
+     */
+    function CustomProjection(url, ellipsoid) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string('url', url);
+        //>>includeEnd('debug');
+
+        this._ellipsoid = defaultValue(ellipsoid, Ellipsoid.WGS84);
+
+        this._project = undefined;
+        this._unproject = undefined;
+
+        var absoluteUrl = getAbsoluteUri(url);
+        this._url = absoluteUrl;
+
+        this._ready = false;
+        this._readyPromise = buildCustomProjection(this, absoluteUrl);
+    }
+
+    defineProperties(CustomProjection.prototype, {
+        /**
+         * Gets the {@link Ellipsoid}.
+         *
+         * @memberof CustomProjection.prototype
+         *
+         * @type {Ellipsoid}
+         * @readonly
+         */
+        ellipsoid : {
+            get : function() {
+                return this._ellipsoid;
+            }
+        },
+
+        /**
+         * Gets whether or not the projection evenly maps meridians to vertical lines.
+         * This is not guaranteed for custom projections and is assumed false.
+         *
+         * @memberof CustomProjection.prototype
+         *
+         * @type {Boolean}
+         * @readonly
+         * @private
+         */
+        isNormalCylindrical : {
+            get : function() {
+                return false;
```

Comment:
Why isn't this part of the custom projection plugin API?  Or at the very least a parameter to the constructor so that it can be set? (Though that seems less useful)

Reply:
This is a flag to tell anything in Cesium that normally depends on the projection being rectangular (like bounding spheres, 2D infinite scroll, terrain tile processing) that the assumption isn't correct and that Cesium should use the code paths with more computation and more vertex data.

If it was part of the API we wouldn't ever want users to change it. At the moment it's not accessible at all due to consequences of abuse. Users might set it to `true` for weird projections because "why not" and break everything, or even make Cesium completely unresponsive.